### PR TITLE
Refactorings

### PR DESCRIPTION
**Pull request by DLochmelis33**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -47,7 +47,7 @@ kotlin {
     sourceSets {
         val commonMain by getting {
             dependencies {
-//                implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.6.4")
+//                implementation("org.jetbrains.kotlin:kotlin-stdlib:1.9.0-RC")
```
Comment (eupp) [inline build.gradle.kts]: please remove commented line

File: README.md [hunk 2]
```diff
@@ -38,7 +40,9 @@ Main function currently shows an example of how to use this project. To write a
...
 ## Notes
 
-* Setting thread affinity is not supported on MacOS (yet?). As such, `getAffinityManager()` returns `null` on MacOS. 
-* Currently, there is no CLI or any other defined interface. The `main()` function is quite messy and likely will remain so (at least for now).
-* For some reason, running a lot of different tests in one go will drastically reduce the performance and weak outcomes frequency. For now, better avoid running for longer than 5 minutes.
+* Setting thread affinity is not supported on macOS (yet?). As such, `getAffinityManager()` returns `null` on macOS.
+* Currently, there is no CLI or any other defined interface. The `main()` function is quite messy and likely will remain
+  so (at least for now).
+* For some reason, running a lot of different tests in one go will drastically reduce the performance and weak outcomes'
+  frequency. For now, better avoid running for longer than 5 minutes.
```
Comment (eupp) [inline README.md]: Are there any substantial changes in `README.md`?
If not I would propose to revert the changes to keep the diff shorter.

File: src/commonMain/kotlin/komem/litmus/LitmusOutcome.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package komem.litmus
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/LitmusOutcome.kt]: please add an empty line after package declaration

File: src/commonMain/kotlin/komem/litmus/LitmusOutcome.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+enum class LitmusOutcomeType { ACCEPTED, INTERESTING, FORBIDDEN }
+
+data class LitmusOutcomeInfo(
+    val outcome: LitmusOutcome,
+    val count: Long,
+    val type: LitmusOutcomeType?,
+)
+
+data class OutcomeSetupScope(
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/LitmusOutcome.kt]: please name `LitmusOutcomeSetupScope` for consistency

File: src/commonMain/kotlin/komem/litmus/LitmusOutcome.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+) {
+    fun getType(outcome: LitmusOutcome) = when (outcome) {
+        in accepted -> LitmusOutcomeType.ACCEPTED
+        in interesting -> LitmusOutcomeType.INTERESTING
+        in forbidden -> LitmusOutcomeType.FORBIDDEN
+        else -> default
+    }
+}
+
+fun List<LitmusOutcome>.groupToInfo(outcomeSetup: OutcomeSetupScope?): List<LitmusOutcomeInfo> = this
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/LitmusOutcome.kt]: I think `groupIntoInfo` is grammatically correct?

File: src/commonMain/kotlin/komem/litmus/LitmusResult.kt [hunk 6]
```diff
@@ -1,7 +1,11 @@
+package komem.litmus
+
+typealias LitmusResult = List<LitmusOutcomeInfo>
+
 val LitmusResult.interestingFrequency: Double
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/LitmusResult.kt]: Is this function used in the project (IDEA says it's unused). 
Please either remove it, or generalize it to accept `LitmusOutcomeType` as a parameter (and rename correspondingly).

File: src/commonMain/kotlin/komem/litmus/MemShuffler.kt [hunk 7]
```diff
@@ -1,8 +1,10 @@
+package komem.litmus
+
 import kotlin.properties.ReadWriteProperty
 import kotlin.random.Random
 import kotlin.reflect.KProperty
 
-class MemShuffler(bufferSize: Int, random: Random = Random) : ReadWriteProperty<BasicLitmusTest, Int> {
+class MemShuffler(bufferSize: Int, random: Random = Random) : ReadWriteProperty<LitmusTest, Int> {
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/MemShuffler.kt]: Please rename into `MemoryShuffler`

File: src/commonMain/kotlin/komem/litmus/RunningParams.kt [hunk 8]
```diff
@@ -0,0 +1,39 @@
+package komem.litmus
+
+import komem.litmus.barriers.BarrierProducer
+
+data class RunningParams(
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/RunningParams.kt]: Please rename into `RunOptions`

File: src/commonMain/kotlin/komem/litmus/RunningParams.kt [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    val batchSize: Int,
+    val syncPeriod: Int,
+    val affinityMap: AffinityMap?,
+    val memShufflerProducer: (() -> MemShuffler)?,
+    val barrierProducer: BarrierProducer,
+)
+
+fun variateParams(
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/RunningParams.kt]: `variateRunOptions`

File: src/commonMain/kotlin/komem/litmus/Utils.kt [hunk 10]
```diff
@@ -1,9 +1,11 @@
+package komem.litmus
+
 import platform.posix._SC_NPROCESSORS_ONLN
 import platform.posix.sysconf
 
 fun List<List<String>>.tableFormat(hasHeader: Boolean = false): String {
     val columnCount = maxOf { it.size }
-    val columnSizes = (0 until columnCount).map { i ->
+    val columnSizes = (0..<columnCount).map { i ->
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/Utils.kt]: please add spaces

File: src/commonMain/kotlin/komem/litmus/statistics/ChiSquareTest.kt [hunk 11]
```diff
@@ -7,7 +9,7 @@ fun chiSquaredTest(sample1: List<Int>, sample2: List<Int>): Boolean {
     val obs = Array(2) { IntArray(n) } // observed frequencies
 
     // fill in observed frequencies
-    for (i in 0 until n) {
+    for (i in 0..<n) {
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/statistics/ChiSquareTest.kt]: forgot spaces (everywhere in this file)

File: src/commonMain/kotlin/komem/litmus/statistics/Dev.kt [hunk 12]
```diff
@@ -0,0 +1,30 @@
+package komem.litmus.statistics
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/statistics/Dev.kt]: What does `Dev` stands for in the name of this file?

File: src/commonMain/kotlin/komem/litmus/tests/KomemTests.kt [hunk 13]
```diff
@@ -183,8 +187,33 @@ class MPTest : BasicLitmusTest("message passing") {
...
+//    thread {
+//        val a = y
+//        val b = x
+//        outcome = a to b
+//    }
+//    outcomes {
+//        accepted = setOf(0 to 0, 0 to 1, 1 to 1)
+//        interesting = setOf(1 to 0)
+//    }
+//}
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/tests/KomemTests.kt]: could you please remove the commented part

File: src/commonMain/kotlin/komem/litmus/tests/KomemTests.kt [hunk 14]
```diff
@@ -23,26 +24,30 @@ class AtomTest : BasicLitmusTest("access atomicity") {
             accepted = setOf(0, -1)
         }
     }
+
+//    @Test
+//    fun test() = this.runTest()
+
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/tests/KomemTests.kt]: please remove

File: src/linuxMain/kotlin/komem/litmus/AffinityBindingsImplPosix.kt [hunk 15]
```diff
@@ -28,9 +30,9 @@ private fun setAffinity(thread: pthread_t, cpus: Set<Int>): Unit = memScoped {
 private fun getAffinity(thread: pthread_t): Set<Int> = memScoped {
     val set = alloc<cpu_set_t>()
     get_affinity(thread, set.ptr).callCheck()
-    return (0 until cpu_setsize())
-            .filter { cpu_isset(it, set.ptr) != 0 }
-            .toSet()
+    return (0..<cpu_setsize())
```
Comment (eupp) [inline src/linuxMain/kotlin/komem/litmus/AffinityBindingsImplPosix.kt]: spacing issues again

File: src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt [hunk 16]
```diff
@@ -0,0 +1,74 @@
+package komem.litmus.runners
+
+import komem.litmus.*
+import kotlin.time.Duration
+import kotlin.time.TimeSource
+
+fun LitmusTestRunner.runTest(
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt]: Why you decided to move these functions into separate file? 
It is perfectly fine to place the extensions functions in the same file where the interface itself is defined (`LitmusTestRunner.kt` in this case).

File: src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    while (start.elapsedNow() < timeLimit) {
+        val outcomes = runTest(params, testProducer)
+        val outcomeSetup = testProducer().getOutcomeSetup()
+        results.addAll(outcomes.groupToInfo(outcomeSetup))
+    }
+    return results.mergeOutcomes()
+}
+
+/*
+ Note: interprets AffinityMap as a sequence of smaller maps
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt]: Block comments should have `*` at the beginning of each line.
https://google.github.io/styleguide/javaguide.html#s4.8.6-comments

File: src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+        val newParams = params.copy(affinityMap = newAffinityMap)
+        runTest(newParams, testProducer)
+    }
+    return allOutcomes.flatten()
+}
+
+fun LitmusTestRunner.runTestParallel(
+    params: RunningParams,
+    testProducer: () -> LitmusTest,
+): List<LitmusOutcome> = with(cpuCount() / testProducer().overriddenThreads().size) {
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt]: `with` is overused here (you use it for `Int`).
`with` is generally used to call a bunch a functions in a context of some object:
https://kotlinlang.org/docs/scope-functions.html#with

Usually there is no point to use `with` with primitive types. 

Besides, the code without `with` is much cleaner here. 

```
val instances = cpuCount() / testProducer().overriddenThreads().size
return runTestParallel(params, instances, testProducer)
```

or 


```
... = runTestParallel(params = params, testProducer = testProducer, instances = cpuCount() / testProducer().overriddenThreads().size)
```

File: src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+/*
+ Note: interprets AffinityMap as a sequence of smaller maps
+ Example: for a map [ [0], [1], [2], [3] ],a test with 2 threads, and 2 instances, the
+ first instance will have a [ [0], [1] ] map and the second one will have [ [2], [3] ].
+ */
+fun LitmusTestRunner.runTestParallel(
+    params: RunningParams,
+    instances: Int,
```
Comment (eupp) [inline src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt]: Similarly, as you make `timeLimit` the first argument in the previous override, it makes sense to make `instances` first argument here too. 

In general, it is a good practice to keep the same order of arguments in all overloads (and do not insert new arguments in-between).

**Review by eupp — CHANGES_REQUESTED** [review]
Great job!

Still a few cosmetics fixes are requires. Please see my comments. 
After they will be resolved it is ready to go.

File: README.md (hunk 2 above)
Comment (DLochmelis33) [reply README.md]: MacOS -> macOS (because IDEA told so)

File: src/commonMain/kotlin/komem/litmus/RunningParams.kt (hunk 8 above)
Comment (DLochmelis33) [reply src/commonMain/kotlin/komem/litmus/RunningParams.kt]: I like "Params" better :) IMO "options" is more about preferences / general settings

File: src/commonMain/kotlin/komem/litmus/Utils.kt (hunk 10 above)
Comment (DLochmelis33) [reply src/commonMain/kotlin/komem/litmus/Utils.kt]: the autoformatter explicitly removes these spaces

File: src/commonMain/kotlin/komem/litmus/RunningParams.kt (hunk 8 above)
Comment (eupp) [reply src/commonMain/kotlin/komem/litmus/RunningParams.kt]: Then `Parameters` is better I guess (since it is not an abbreviation). 
This is why I prefer `Options` --- this word is shorter and not an abbreviation.

Comment (eupp) [reply src/commonMain/kotlin/komem/litmus/RunningParams.kt]: Ok, if you prefere "parameters", let's keep them, but I would recommend then to rename to full form `Parameters`.

File: src/commonMain/kotlin/komem/litmus/Utils.kt (hunk 10 above)
Comment (eupp) [reply src/commonMain/kotlin/komem/litmus/Utils.kt]: Then disable this behavior in the IDEA settings ))

Comment (eupp) [reply src/commonMain/kotlin/komem/litmus/Utils.kt]: Ok, it looks like official styleguide actually recommends to remove spaces:
https://kotlinlang.org/docs/coding-conventions.html#horizontal-whitespace

But I don't know, looks weird for me (harder to read), harder to refactor (in case when you decide to replace it to e.g. `downTo`), inconsistent with the rules for other binary expressions.

File: src/commonMain/kotlin/komem/litmus/statistics/ChiSquareTest.kt (hunk 11 above)
Comment (DLochmelis33) [reply src/commonMain/kotlin/komem/litmus/statistics/ChiSquareTest.kt]: autoformatter again

File: src/commonMain/kotlin/komem/litmus/statistics/Dev.kt (hunk 12 above)
Comment (DLochmelis33) [reply src/commonMain/kotlin/komem/litmus/statistics/Dev.kt]: deleted this file

File: src/linuxMain/kotlin/komem/litmus/AffinityBindingsImplPosix.kt (hunk 15 above)
Comment (DLochmelis33) [reply src/linuxMain/kotlin/komem/litmus/AffinityBindingsImplPosix.kt]: autoformatter again

File: src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt (hunk 16 above)
Comment (DLochmelis33) [reply src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt]: Moved back

File: src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt (hunk 18 above)
Comment (DLochmelis33) [reply src/commonMain/kotlin/komem/litmus/runners/LitmusTestRunnerUtil.kt]: I saved one line though! Fine, fixed

**Comment by DLochmelis33** [comment]
I messed up force-pushing, but this PR is in fact successfully merged
